src/Shared/FileUtilities.cs

```diff
@@ -165,6 +165,30 @@ internal static int GetPathsHash(IEnumerable<string> assemblyPaths)
             return builder.ToString().GetHashCode();
         }
 
+        /// <summary>
+        /// Returns whether MSBuild can write to the given directory. Throws for PathTooLongExceptions
+        /// but not other exceptions.
+        /// </summary>
+        internal static bool CanWriteToDirectory(string directory)
+        {
+            try
+            {
+                string testFilePath = Path.Combine(directory, "textFile.txt");
+                File.WriteAllText(testFilePath, "Successfully wrote to file.");
```

Comment:
With some tweaks it'll be much easier to determine if this is causing problems, and support concurrent execution:

``suggestion
                string testFilePath = Path.Combine(directory, $"MSBuild_{Guid.NewGuid().ToString("N")}.txt");                File.WriteAllText(testFilePath, $"MSBuild process {Process.GetCurrentProcess().Id} successfully wrote to file.");
```